File: src/main/java/io/github/oleksivio/telegram/bot/core/controller/processor/session/SessionProcessor.java
package io.github.oleksivio.telegram.bot.core.controller.processor.session;

import io.github.oleksivio.telegram.bot.api.model.objects.Update;
import io.github.oleksivio.telegram.bot.api.model.result.HandlerResult;
import org.springframework.stereotype.Controller;

import java.util.ArrayList;
import java.util.Collection;

@Controller
public class SessionProcessor {

    private final Collection<SessionManager> sessionManagers = new ArrayList<>();

    public HandlerResult receive(Update update) {

        // check already activated sessions
        for (SessionManager manager : sessionManagers) {
            if (manager.isActive(update)) {
                HandlerResult handlerResult = manager.receive(update);
                if (!handlerResult.isPassed()) {
                    return handlerResult;
                }
            }
        }

        // if no any active session receive this update
        // find in initial method
        for (SessionManager manager : sessionManagers) {
            HandlerResult handlerResult = manager.init(update);
            if (handlerResult.hasSuccess()) {
                return handlerResult;
            }
        }

        return HandlerResult.pass();
    }


    public void add(SessionManager sessionManager) {
        sessionManagers.add(sessionManager);
    }
}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/core/controller/processor/session/SessionProcessor.kt
package io.github.oleksivio.telegram.bot.core.controller.processor.session

import io.github.oleksivio.telegram.bot.api.model.objects.Update
import io.github.oleksivio.telegram.bot.api.model.result.HandlerResult
import org.springframework.stereotype.Controller
import java.util.*

@Controller
class SessionProcessor {

    private val sessionManagers = ArrayList<SessionManager>()

    fun receive(update: Update): HandlerResult {

        // check already activated sessions
        for (manager in sessionManagers) {
            if (manager.isActive(update)) {
                val handlerResult = manager.receive(update)
                if (!handlerResult.isPassed) {
                    return handlerResult
                }
            }
        }

        // if no any active session receive this update
        // find in initial method
        for (manager in sessionManagers) {
            val handlerResult = manager.init(update)
            if (handlerResult.hasSuccess()) {
                return handlerResult
            }
        }

        return HandlerResult.pass()
    }

    fun add(sessionManager: SessionManager) {
        sessionManagers.add(sessionManager)
    }
}
